storage/index/block.go

```diff
@@ -0,0 +1,377 @@
+	if err != nil {
+		return nil, err
+	}
+
+	b := &block{
+		state:   blockStateOpen,
+		segment: seg,
+
+		startTime: startTime,
+		endTime:   startTime.Add(blockSize),
+		opts:      opts,
+		logger:    opts.InstrumentOptions().Logger(),
+	}
+	b.newExecutorFn = b.executorWithRLock
+
+	return b, nil
+}
+
+func (b *block) StartTime() time.Time {
+	return b.startTime
+}
+
+func (b *block) EndTime() time.Time {
+	return b.endTime
+}
+
+func (b *block) WriteBatch(inserts []WriteBatchEntry) (WriteBatchResult, error) {
+	b.RLock()
+	defer b.RUnlock()
+	if b.state != blockStateOpen {
+		// NB: releasing all references to inserts
+		for _, insert := range inserts {
+			insert.OnIndexSeries.OnIndexFinalize()
+		}
+		return WriteBatchResult{
+			NumError: int64(len(inserts)),
+		}, writeBatchErrorInvalidState(b.state)
+	}
+
+	// FOLLOWUP(prateek): docs array pooling
+	batch := m3ninxindex.Batch{
+		Docs:                make([]doc.Document, 0, len(inserts)),
+		AllowPartialUpdates: true,
+	}
+	for _, insert := range inserts {
+		batch.Docs = append(batch.Docs, insert.Document)
+	}
+
+	err := b.segment.InsertBatch(batch)
+	if err == nil {
+		for idx, insert := range inserts {
+			insert.OnIndexSeries.OnIndexSuccess(b.endTime)
+			insert.OnIndexSeries.OnIndexFinalize()
+			batch.Docs[idx] = doc.Document{}
+		}
+		return WriteBatchResult{
+			NumSuccess: int64(len(inserts)),
+		}, nil
+	}
+
+	partialErr, ok := err.(*m3ninxindex.BatchPartialError)
+	if !ok {
+		// should never happen
+		b.logger.Errorf(
+			"[invariant violated] received non BatchPartialError from m3ninx InsertBatch [%T]", err)
+		// NB: marking all the inserts as failure, cause we don't know which ones failed
+		for _, insert := range inserts {
+			insert.OnIndexSeries.OnIndexFinalize()
+			insert.Document = doc.Document{}
+			insert.OnIndexSeries = nil
+		}
+		return WriteBatchResult{
+			NumError: int64(len(inserts)),
+		}, fmt.Errorf("unexpected error: %v", err)
+	}
+
+	// first finalize all the responses which were errors, and mark them
+	// nil to indicate they're done.
+	numErr := len(partialErr.Indices())
+	for _, idx := range partialErr.Indices() {
+		inserts[idx].OnIndexSeries.OnIndexFinalize()
+		inserts[idx].OnIndexSeries = nil
+		inserts[idx].Document = doc.Document{}
+	}
+
+	// mark all non-error inserts success, so we don't repeatedly index them,
+	// and then finalize any held references.
+	for _, insert := range inserts {
+		if insert.OnIndexSeries == nil {
+			continue
+		}
+		insert.OnIndexSeries.OnIndexSuccess(b.endTime)
+		insert.OnIndexSeries.OnIndexFinalize()
+		insert.OnIndexSeries = nil
+		insert.Document = doc.Document{}
+	}
+
+	return WriteBatchResult{
+		NumSuccess: int64(len(inserts) - numErr),
+		NumError:   int64(numErr),
+	}, partialErr
+}
+
+func (b *block) executorWithRLock() (search.Executor, error) {
+	var (
+		readers = make([]m3ninxindex.Reader, 0, 1+len(b.bootstrappedSegments))
+		success = false
+	)
+
+	// cleanup in case shit breaks
```

Comment:
super nit: Obviously not a big deal, but maybe `s/shit/all the things/`?

Reply:
sure thing

Reply:
done